sdl_android/src/main/java/com/smartdevicelink/managers/SdlManager.java

```diff
@@ -129,94 +127,137 @@ public void onError(String info, Exception e){
+				fileManager.uploadArtwork(SdlManager.this.appIcon, new CompletionListener() {
+					@Override
+					public void onComplete(boolean success) {
+						if (success) {
+							SetAppIcon msg = new SetAppIcon(SdlManager.this.appIcon.getName());
+							_internalInterface.sendRPCRequest(msg);
+						}
 					}
-				}
+				});
+			} else {
+				SetAppIcon msg = new SetAppIcon(SdlManager.this.appIcon.getName());
+				_internalInterface.sendRPCRequest(msg);
 			}
 		}
-	};
+	}
 
 	protected void initialize(){
-		// instantiate managers
-
+		// Instantiate sub managers
 		this.permissionManager = new PermissionManager(_internalInterface);
-		this.permissionManager.start(subManagerListener);
-
 		this.fileManager = new FileManager(_internalInterface, context);
-		this.fileManager.start(subManagerListener);
-
 		if (lockScreenConfig.isEnabled()) {
 			this.lockScreenManager = new LockScreenManager(lockScreenConfig, context, _internalInterface);
-			this.lockScreenManager.start(subManagerListener);
 		}
-
 		this.screenManager = new ScreenManager(_internalInterface, this.fileManager);
-		this.screenManager.start(subManagerListener);
-
-
 		if(getAppTypes().contains(AppHMIType.NAVIGATION) || getAppTypes().contains(AppHMIType.PROJECTION)){
 			this.videoStreamingManager = new VideoStreamingManager(_internalInterface);
-		}else{
+		} else {
 			this.videoStreamingManager = null;
 		}
-
-
 		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN
 				&& (getAppTypes().contains(AppHMIType.NAVIGATION) || getAppTypes().contains(AppHMIType.PROJECTION)) ) {
 			this.audioStreamManager = new AudioStreamManager(_internalInterface, context);
 		} else {
 			this.audioStreamManager = null;
 		}
 
+		// Start sub managers
+		this.permissionManager.start(subManagerListener);
+		this.fileManager.start(subManagerListener);
+		if (lockScreenConfig.isEnabled()){
+			this.lockScreenManager.start(subManagerListener);
+		}
+		this.screenManager.start(subManagerListener);
+	}
+
+	public int getState() {
+		synchronized (STATE_LOCK) {
+			return state;
+		}
+	}
+
+	protected void transitionToState(int state) {
+		synchronized (STATE_LOCK) {
+			this.state = state;
+		}
 	}
 
 	public void dispose() {
-		this.permissionManager.dispose();
+		if (this.permissionManager != null) {
+			this.permissionManager.dispose();
+		}
 
-		this.fileManager.dispose();
+		if (this.fileManager != null) {
+			this.fileManager.dispose();
+		}
 
 		if (this.lockScreenManager != null) {
 			this.lockScreenManager.dispose();
 		}
 
-		this.screenManager.dispose();
+		if (this.screenManager != null) {
+			this.screenManager.dispose();
+		}
 
 		if(this.videoStreamingManager != null) {
 			this.videoStreamingManager.dispose();
 		}
 
-		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
-			if (this.audioStreamManager != null) {
-				this.audioStreamManager.dispose();
-			}
+		if (this.audioStreamManager != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
```

Comment:
Honestly we probably don't need the version check here. If the `AudioStreamManager` is not null it got created because this check has already passed and if it is not null it should be `disposed`